File: loan_calc.go
// loan_calc.go - Кредитный калькулятор на Go (CLI)
package main

import (
	"bufio"
	"encoding/csv"
	"flag"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

type LoanCalculator struct {
	amount      float64
	annualRate  float64
	term        int
	paymentType string
	monthlyRate float64
}

type PaymentRow struct {
	Month     int
	Payment   float64
	Principal float64
	Interest  float64
	Balance   float64
}

func NewLoanCalculator(amount float64, rate float64, term int, ptype string) *LoanCalculator {
	return &LoanCalculator{
		amount:      amount,
		annualRate:  rate,
		term:        term,
		paymentType: ptype,
		monthlyRate: rate / 100 / 12,
	}
}

func (lc *LoanCalculator) Calculate() ([]PaymentRow, float64, float64) {
	schedule := []PaymentRow{}
	balance := lc.amount
	totalPaid := 0.0
	var payment, principal, interest float64

	if lc.paymentType == "annuity" {
		if lc.monthlyRate == 0 {
			payment = lc.amount / float64(lc.term)
		} else {
			pow := math.Pow(1+lc.monthlyRate, float64(lc.term))
			payment = lc.amount * lc.monthlyRate * pow / (pow - 1)
		}
		for month := 1; month <= lc.term; month++ {
			interest = balance * lc.monthlyRate
			principal = payment - interest
			if principal > balance {
				principal = balance
				payment = principal + interest
			}
			balance -= principal
			totalPaid += payment
			schedule = append(schedule, PaymentRow{
				Month:     month,
				Payment:   payment,
				Principal: principal,
				Interest:  interest,
				Balance:   balance,
			})
		}
	} else { // diff
		principalPerMonth := lc.amount / float64(lc.term)
		for month := 1; month <= lc.term; month++ {
			interest = balance * lc.monthlyRate
			payment = principalPerMonth + interest
			if payment > balance+interest {
				payment = balance + interest
				principalPerMonth = balance
			}
			balance -= principalPerMonth
			totalPaid += payment
			schedule = append(schedule, PaymentRow{
				Month:     month,
				Payment:   payment,
				Principal: principalPerMonth,
				Interest:  interest,
				Balance:   balance,
			})
		}
	}
	overpayment := totalPaid - lc.amount
	return schedule, totalPaid, overpayment
}

func (lc *LoanCalculator) PrintSchedule(schedule []PaymentRow) {
	fmt.Println("\nГрафик платежей:")
	fmt.Printf("%-4s %-12s %-15s %-12s %-12s\n", "№", "Платёж", "Основной долг", "Проценты", "Остаток")
	for _, r := range schedule {
		fmt.Printf("%-4d %10.2f %14.2f %11.2f %11.2f\n", r.Month, r.Payment, r.Principal, r.Interest, r.Balance)
	}
}

func (lc *LoanCalculator) ExportCSV(filepath string, schedule []PaymentRow) {
	file, err := os.Create(filepath)
	if err != nil {
		fmt.Println("Ошибка создания файла:", err)
		return
	}
	defer file.Close()
	writer := csv.NewWriter(file)
	defer writer.Flush()
	writer.Write([]string{"month", "payment", "principal", "interest", "balance"})
	for _, r := range schedule {
		writer.Write([]string{
			strconv.Itoa(r.Month),
			fmt.Sprintf("%.2f", r.Payment),
			fmt.Sprintf("%.2f", r.Principal),
			fmt.Sprintf("%.2f", r.Interest),
			fmt.Sprintf("%.2f", r.Balance),
		})
	}
}

func main() {
	var (
		amount  float64
		rate    float64
		term    int
		ptype   string
		scheduleFlag bool
		output  string
	)
	flag.Float64Var(&amount, "amount", 0, "Сумма кредита")
	flag.Float64Var(&rate, "rate", 0, "Годовая ставка (%)")
	flag.IntVar(&term, "term", 0, "Срок (месяцев)")
	flag.StringVar(&ptype, "type", "annuity", "Тип платежа (annuity/diff)")
	flag.BoolVar(&scheduleFlag, "schedule", false, "Показать график")
	flag.StringVar(&output, "output", "", "Сохранить график в CSV")
	flag.Parse()

	if amount == 0 || rate == 0 || term == 0 {
		interactiveMode()
		return
	}

	calc := NewLoanCalculator(amount, rate, term, ptype)
	schedule, total, overpayment := calc.Calculate()
	fmt.Printf("\nСумма кредита: %.2f\n", amount)
	fmt.Printf("Ставка: %.2f%% годовых\n", rate)
	fmt.Printf("Срок: %d месяцев\n", term)
	fmt.Printf("Тип платежа: %s\n", map[string]string{"annuity": "Аннуитетный", "diff": "Дифференцированный"}[ptype])
	fmt.Printf("Общая сумма выплат: %.2f\n", total)
	fmt.Printf("Переплата: %.2f\n", overpayment)
	if scheduleFlag {
		calc.PrintSchedule(schedule)
	}
	if output != "" {
		calc.ExportCSV(output, schedule)
		fmt.Printf("График сохранён в %s\n", output)
	}
}

func interactiveMode() {
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Println("Интерактивный режим кредитного калькулятора")
	fmt.Print("Сумма кредита: ")
	scanner.Scan()
	amount, _ := strconv.ParseFloat(scanner.Text(), 64)
	fmt.Print("Годовая ставка (%): ")
	scanner.Scan()
	rate, _ := strconv.ParseFloat(scanner.Text(), 64)
	fmt.Print("Срок (месяцев): ")
	scanner.Scan()
	term, _ := strconv.Atoi(scanner.Text())
	fmt.Print("Тип платежа (annuity/diff): ")
	scanner.Scan()
	ptype := strings.ToLower(scanner.Text())
	if ptype != "annuity" && ptype != "diff" {
		ptype = "annuity"
	}
	calc := NewLoanCalculator(amount, rate, term, ptype)
	schedule, total, overpayment := calc.Calculate()
	fmt.Printf("\nСумма кредита: %.2f\n", amount)
	fmt.Printf("Ставка: %.2f%%\n", rate)
	fmt.Printf("Срок: %d мес.\n", term)
	fmt.Printf("Тип: %s\n", map[string]string{"annuity": "Аннуитетный", "diff": "Дифференцированный"}[ptype])
	fmt.Printf("Общая сумма выплат: %.2f\n", total)
	fmt.Printf("Переплата: %.2f\n", overpayment)
	fmt.Print("Показать график платежей? (y/n): ")
	scanner.Scan()
	if strings.ToLower(scanner.Text()) == "y" {
		calc.PrintSchedule(schedule)
	}
	fmt.Print("Сохранить график в CSV? (y/n): ")
	scanner.Scan()
	if strings.ToLower(scanner.Text()) == "y" {
		fmt.Print("Имя файла (по умолчанию schedule.csv): ")
		scanner.Scan()
		fname := scanner.Text()
		if fname == "" {
			fname = "schedule.csv"
		}
		calc.ExportCSV(fname, schedule)
		fmt.Printf("Сохранено в %s\n", fname)
	}
}

File: loan_calc.py
#!/usr/bin/env python3
"""
loan_calc.py - Кредитный калькулятор на Python (CLI + Tkinter GUI)
Поддерживает: аннуитетный и дифференцированный платежи, график, экспорт CSV.
"""
import argparse
import sys
import csv
import os
from math import pow
from datetime import datetime
from typing import List, Tuple, Dict

try:
    import tkinter as tk
    from tkinter import ttk, filedialog, messagebox, scrolledtext
    GUI_AVAILABLE = True
except ImportError:
    GUI_AVAILABLE = False

class LoanCalculator:
    def __init__(self, amount: float, annual_rate: float, term_months: int, payment_type: str):
        self.amount = amount
        self.annual_rate = annual_rate
        self.term = term_months
        self.payment_type = payment_type.lower()  # 'annuity' or 'diff'
        self.monthly_rate = annual_rate / 100 / 12

    def calculate(self) -> Tuple[List[Dict], float, float]:
        """
        Возвращает (график_платежей, общая_сумма_выплат, переплата)
        График: список словарей с ключами: month, payment, principal, interest, balance
        """
        schedule = []
        balance = self.amount
        total_paid = 0.0
        if self.payment_type == 'annuity':
            if self.monthly_rate == 0:
                payment = self.amount / self.term
            else:
                payment = self.amount * self.monthly_rate * pow(1 + self.monthly_rate, self.term) / (pow(1 + self.monthly_rate, self.term) - 1)
            for month in range(1, self.term + 1):
                interest = balance * self.monthly_rate
                principal = payment - interest
                if principal > balance:
                    principal = balance
                    payment = principal + interest
                balance -= principal
                total_paid += payment
                schedule.append({
                    'month': month,
                    'payment': payment,
                    'principal': principal,
                    'interest': interest,
                    'balance': balance
                })
        else:  # дифференцированный
            principal_per_month = self.amount / self.term
            for month in range(1, self.term + 1):
                interest = balance * self.monthly_rate
                payment = principal_per_month + interest
                if payment > balance + interest:
                    payment = balance + interest
                    principal_per_month = balance
                balance -= principal_per_month
                total_paid += payment
                schedule.append({
                    'month': month,
                    'payment': payment,
                    'principal': principal_per_month,
                    'interest': interest,
                    'balance': balance
                })
        overpayment = total_paid - self.amount
        return schedule, total_paid, overpayment

    def export_csv(self, filepath: str, schedule: List[Dict]):
        with open(filepath, 'w', newline='', encoding='utf-8') as f:
            writer = csv.DictWriter(f, fieldnames=['month', 'payment', 'principal', 'interest', 'balance'])
            writer.writeheader()
            writer.writerows(schedule)

    def print_schedule(self, schedule: List[Dict]):
        print("\nГрафик платежей:")
        print(f"{'№':<4} {'Платёж':<12} {'Основной долг':<15} {'Проценты':<12} {'Остаток':<12}")
        for row in schedule:
            print(f"{row['month']:<4} {row['payment']:>10.2f} {row['principal']:>14.2f} {row['interest']:>11.2f} {row['balance']:>11.2f}")

# ========== CLI ==========
def cli():
    parser = argparse.ArgumentParser(description="Кредитный калькулятор")
    parser.add_argument("--amount", type=float, help="Сумма кредита")
    parser.add_argument("--rate", type=float, help="Годовая ставка (%)")
    parser.add_argument("--term", type=int, help="Срок в месяцах")
    parser.add_argument("--type", choices=['annuity','diff'], default='annuity', help="Тип платежа")
    parser.add_argument("--schedule", action="store_true", help="Показать график платежей")
    parser.add_argument("--output", help="Сохранить график в CSV")
    parser.add_argument("--gui", action="store_true", help="Запустить GUI")
    args = parser.parse_args()

    if args.gui and GUI_AVAILABLE:
        root = tk.Tk()
        app = LoanCalculatorGUI(root)
        root.mainloop()
        return

    if args.amount is None or args.rate is None or args.term is None:
        interactive_mode()
        return

    calc = LoanCalculator(args.amount, args.rate, args.term, args.type)
    schedule, total, overpayment = calc.calculate()
    print(f"\nСумма кредита: {args.amount:.2f}")
    print(f"Ставка: {args.rate:.2f}% годовых")
    print(f"Срок: {args.term} месяцев")
    print(f"Тип платежа: {'Аннуитетный' if args.type == 'annuity' else 'Дифференцированный'}")
    print(f"Общая сумма выплат: {total:.2f}")
    print(f"Переплата: {overpayment:.2f}")
    if args.schedule:
        calc.print_schedule(schedule)
    if args.output:
        calc.export_csv(args.output, schedule)
        print(f"График сохранён в {args.output}")

def interactive_mode():
    print("Интерактивный режим кредитного калькулятора")
    try:
        amount = float(input("Сумма кредита: "))
        rate = float(input("Годовая ставка (%): "))
        term = int(input("Срок (месяцев): "))
        ptype = input("Тип платежа (annuity/diff): ").strip().lower()
        if ptype not in ['annuity','diff']:
            ptype = 'annuity'
        calc = LoanCalculator(amount, rate, term, ptype)
        schedule, total, overpayment = calc.calculate()
        print(f"\nСумма кредита: {amount:.2f}")
        print(f"Ставка: {rate:.2f}%")
        print(f"Срок: {term} мес.")
        print(f"Тип: {'Аннуитетный' if ptype == 'annuity' else 'Дифференцированный'}")
        print(f"Общая сумма выплат: {total:.2f}")
        print(f"Переплата: {overpayment:.2f}")
        show = input("Показать график платежей? (y/n): ").strip().lower()
        if show == 'y':
            calc.print_schedule(schedule)
        save = input("Сохранить график в CSV? (y/n): ").strip().lower()
        if save == 'y':
            fname = input("Имя файла (по умолчанию schedule.csv): ").strip()
            if not fname:
                fname = "schedule.csv"
            calc.export_csv(fname, schedule)
            print(f"Сохранено в {fname}")
    except Exception as e:
        print(f"Ошибка: {e}")

# ========== GUI ==========
if GUI_AVAILABLE:
    class LoanCalculatorGUI:
        def __init__(self, root):
            self.root = root
            self.root.title("Кредитный калькулятор")
            self.root.geometry("700x600")
            self.root.resizable(True, True)
            self.create_widgets()

        def create_widgets(self):
            main = ttk.Frame(self.root, padding="10")
            main.pack(fill=tk.BOTH, expand=True)

            # Поля ввода
            row = 0
            ttk.Label(main, text="Сумма кредита:").grid(row=row, column=0, sticky="w", pady=5)
            self.amount_var = tk.StringVar()
            ttk.Entry(main, textvariable=self.amount_var).grid(row=row, column=1, padx=5, pady=5)
            row += 1
            ttk.Label(main, text="Годовая ставка (%):").grid(row=row, column=0, sticky="w", pady=5)
            self.rate_var = tk.StringVar()
            ttk.Entry(main, textvariable=self.rate_var).grid(row=row, column=1, padx=5, pady=5)
            row += 1
            ttk.Label(main, text="Срок (месяцев):").grid(row=row, column=0, sticky="w", pady=5)
            self.term_var = tk.StringVar()
            ttk.Entry(main, textvariable=self.term_var).grid(row=row, column=1, padx=5, pady=5)
            row += 1
            ttk.Label(main, text="Тип платежа:").grid(row=row, column=0, sticky="w", pady=5)
            self.type_var = tk.StringVar(value="annuity")
            ttk.Combobox(main, textvariable=self.type_var, values=["annuity", "diff"], state="readonly").grid(row=row, column=1, padx=5, pady=5)
            row += 1
            ttk.Button(main, text="Рассчитать", command=self.calculate).grid(row=row, column=0, columnspan=2, pady=10)
            row += 1
            self.result_text = scrolledtext.ScrolledText(main, height=15, width=70)
            self.result_text.grid(row=row, column=0, columnspan=2, pady=5)
            row += 1
            btn_frame = ttk.Frame(main)
            btn_frame.grid(row=row, column=0, columnspan=2, pady=5)
            ttk.Button(btn_frame, text="Сохранить CSV", command=self.save_csv).pack(side=tk.LEFT, padx=5)
            ttk.Button(btn_frame, text="Очистить", command=self.clear).pack(side=tk.LEFT, padx=5)

        def calculate(self):
            try:
                amount = float(self.amount_var.get())
                rate = float(self.rate_var.get())
                term = int(self.term_var.get())
                ptype = self.type_var.get()
                calc = LoanCalculator(amount, rate, term, ptype)
                schedule, total, overpayment = calc.calculate()
                self.result_text.delete(1.0, tk.END)
                self.result_text.insert(tk.END, f"Сумма кредита: {amount:.2f}\n")
                self.result_text.insert(tk.END, f"Ставка: {rate:.2f}% годовых\n")
                self.result_text.insert(tk.END, f"Срок: {term} месяцев\n")
                self.result_text.insert(tk.END, f"Тип: {'Аннуитетный' if ptype == 'annuity' else 'Дифференцированный'}\n")
                self.result_text.insert(tk.END, f"Общая сумма выплат: {total:.2f}\n")
                self.result_text.insert(tk.END, f"Переплата: {overpayment:.2f}\n\n")
                self.result_text.insert(tk.END, "График платежей:\n")
                self.result_text.insert(tk.END, f"{'№':<4} {'Платёж':<12} {'Основной долг':<15} {'Проценты':<12} {'Остаток':<12}\n")
                for row in schedule:
                    self.result_text.insert(tk.END,
                        f"{row['month']:<4} {row['payment']:>10.2f} {row['principal']:>14.2f} {row['interest']:>11.2f} {row['balance']:>11.2f}\n")
                self._schedule = schedule
            except Exception as e:
                messagebox.showerror("Ошибка", str(e))

        def save_csv(self):
            if hasattr(self, '_schedule'):
                filepath = filedialog.asksaveasfilename(defaultextension=".csv", filetypes=[("CSV", "*.csv")])
                if filepath:
                    calc = LoanCalculator(0,0,0,'annuity')  # dummy
                    calc.export_csv(filepath, self._schedule)
                    messagebox.showinfo("Сохранено", f"Сохранено в {filepath}")
            else:
                messagebox.showwarning("Нет данных", "Сначала выполните расчёт")

        def clear(self):
            self.amount_var.set("")
            self.rate_var.set("")
            self.term_var.set("")
            self.result_text.delete(1.0, tk.END)

if __name__ == "__main__":
    cli()
